Fix recorder Oracle DB models

mciupak opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Three new classes/tables were added to recorder models, however primary key columns 'id' are were added without 'Identity()' parameter. Lack of Identity column breaks recorder functionality on Oracle DB.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: N/A

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mciupak committed:
Fix recorder models for Oracle DB

mciupak committed:
Fix StatisticsRuns

lazdavila reviewed (approved):
Looks correct

emontnemery commented:
@mciupak Do we need to add a migration for existing Oracle databases where these tables were created without the `Identity()`?

mciupak commented:
> @mciupak Do we need to add a migration step for existing Oracle databases where these tables were created without the `Identity()`?

Hmm, tables are already created and cannot be modified (Oracle internals) They have to be deleted and recreated with Identity(), so I think such migration would be useful. Otherwise it would have to be done manually.

emontnemery commented:
OK, take a look at migration for version 18. I think you can copy that but drop + create the StatisticsRuns table too.

mciupak committed:
Update migration for Oracle Identity columns.

emontnemery commented on homeassistant/components/recorder/migration.py:
```diff
@@ -518,6 +518,29 @@ def _apply_update(engine, session, new_version, old_version):  # noqa: C901
+        # Order matters! Statistics has a relation with StatisticsMeta,
+        # so statistics need to be deleted before meta (or in pair depending
+        # on the SQL backend); and meta needs to be created before statistics.
+        if (
```
We only want to do this if the engine dialect is oracle.

mciupak replied:
What about other engines which support Identity colums e.g. Postrges, MSSQL? The state of database will not reflect what is in models.py, or did I miss something?

emontnemery replied:
You mention in the PR description that:
> Lack of Identity column breaks recorder functionality on Oracle DB

There have been no complaints saying that the recorder functionality is broken for any other database engines.

We should only drop tables if they are currently not working because the `Identity()` was missing and there's no way to update the tables to fix that.

emontnemery replied:
Can you clarify what the difference between these options for defining a column are:
`Column(Integer, Identity(), primary_key=True)`
`Column(Integer, primary_key=True)`

The SQLAlchemy documentation is not very clear, but the way I understand it, `Identity()` is optional and can be used to override the default behavior of a primary key?

mciupak replied:
I'm sorry I was wrong.

I did some tests and for DB engines other than Oracle there is not much change how tables are created with or without Identity column e.g. for MySQL AUTO_INCREMENT will be added with or without Identity parameter, for PostgreSQL 'Default' field will be different i.e. without Identity() it will be "nextval('table_id_seq'::regclass)", with Identity() it will be "generated by default as identity", but under the hood the same sequence will be created "nextval('table_id_seq'::regclass)" and so on, so in fact it affects Oracle DB only.

Sorry for any inconveniences, I will update migration logic to be applied for engine dialect oracle only.

emontnemery replied:
No problem at all, thanks a lot for double checking 👍

mciupak commented:
> OK, take a look at migration for version 18. I think you can copy that but drop + create the StatisticsRuns table too.

Alright, I bumped schema version, added requested migration and done some testing. Looks good, except that I had to change migration to schema version 20 and removed oracle from that change as it was failing. Float in Oracle database is by default double precision so no need to change anything there. I tested migration on database with schema version 18 and all went good so I believe we are good to go.

mciupak committed:
Update migration logic

Migration to schema version 22 done only for engine dialect oracle

mciupak commented:
Migration logic to schema version 22 updated to be triggered only for engine dialect oracle.

mciupak committed:
Add missing table check in schema 22 migration
